synapse/federation/transaction_queue.py

```diff
@@ -714,3 +747,183 @@ def json_data_cb():
             success = False
 
         defer.returnValue(success)
+
+
+@attr.s
+class PerDestinationQueue(object):
+    """
+    Holds the per-destination transmission queues.
+
+    (Currently it only holds the EDUs but at some point it would be good to extend it to
+    hold the other per-destination data like PDUs and presence.)
+    """
+    clock = attr.ib()  # type: synapse.util.Clock
+
+    # the callback to be called when our EDUs become ready for transmission.
+    attempt_transaction_cb = attr.ib()  # type: callable
+
+    # edu buckets for this destination, keyed by edu bucket ID
+    edu_buckets = attr.ib(factory=dict)   # type: dict[int, EduTransmissionBucket]
+
+    def get_edu_count(self):
+        """Get the number of pending EDUs for this destination
+
+        Returns:
+            int: number of pending EDUs
+        """
+        return sum(b.get_edu_count() for b in self.edu_buckets.values())
+
+    def is_empty(self):
+        """Check if this queue is empty (and can therefore be destroyed)
+
+        Returns:
+            bool: True if the queue is empty
+        """
+        # it's assumed that we never have any empty buckets, so we just need to
+        # check if we have any buckets at all
+        return not self.edu_buckets
+
+    def get_or_create_edu_bucket(self, bucket_id):
+        """Get (or create) an EDU bucket for the given bucket identifier
+
+        Args:
+            bucket_id (int|None): bucket identifier; None implies instant
+        """
+        if bucket_id is None:
+            bucket_id = INSTANT_EDU_BUCKET_ID
+
+        bucket = self.edu_buckets.get(bucket_id)
+        if bucket:
+            return bucket
+
+        if bucket_id == INSTANT_EDU_BUCKET_ID:
+            delay = 0
+        elif bucket_id == DELAYED_EDU_BUCKET_ID:
+            delay = 5000
+        else:
+            raise RuntimeError("Unknown bucket id %i" % (bucket_id,))
+
+        transmit_task = self.clock.call_later(delay, self.attempt_transaction_cb)
+        transmission_time = self.clock.time_msec() + delay
+        self.edu_buckets[bucket_id] = bucket = EduTransmissionBucket(
+            bucket_transmission_time=transmission_time,
+            transmission_task=transmit_task,
+        )
```

Comment:
Given all the management of the task happens in `PerDestinationQueue` it feels odd for it to be added to the `EduTransmissionBucket` class.